### PR TITLE
{Compute} Fix ValueError in vm commands

### PR DESCRIPTION
**Pull request by houk-ms**
**Description**  

This PR works as a part of fixing the none-actionable issues in #14986

- It fixes the ValueErrors with patterns below in VM commands.

```
not enough values to unpack (expected 4, got 2)
too many values to unpack (expected 4)
...
```
- It applies the AzCLIError in comand group
- It inroduces the new AzCLIErrorType: UsageError, ResourceNotFound 

**Testing Guide**  

**Case1**: Run `az vm image show --urn a:b:c`

before
```
UnexpectedError: The command failed with an unexpected error. Here is the traceback:
not enough values to unpack (expected 4, got 3)
Traceback (most recent call last):
  File "C:\Users\houk\Desktop\msws\env_new\lib\site-packages\knack\cli.py", line 215, in invoke
    cmd_result = self.invocation.execute(args)
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 654, in execute
    raise ex
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 718, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 711, in _run_job
    six.reraise(*sys.exc_info())
  File "C:\Users\houk\Desktop\msws\env_new\lib\site-packages\six.py", line 703, in reraise
    raise value
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 688, in _run_job
    result = cmd_copy(params)
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\__init__.py", line 325, in __call__
    return self.handler(*args, **kwargs)
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\arm.py", line 756, in handler
    show_exception_handler(ex)
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\arm.py", line 772, in show_exception_handler
    raise ex
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli-core\azure\cli\core\commands\arm.py", line 754, in handler
    return getter(**args)
  File "c:\users\houk\desktop\msws\azure-cli\src\azure-cli\azure\cli\command_modules\vm\custom.py", line 1728, in show_vm_image
    publisher, offer, sku, version = urn.split(":")
ValueError: not enough values to unpack (expected 4, got 3)
```
After
```
UsageError: --urn should be in the format of publisher:offer:sku:version
```

**Case2**: Run `az vm image show --urn a:b:c -p d`
Before
```
ValidationError: usage error: --plan STRING --offer STRING --publish STRING --version STRING | --urn STRING
```
After
```
UsageError: If using --urn, do not specify any of --publisher, --offer, --sku, --version
Try to use --urn publisher:offer:sku:version only
```

### CODE REVIEW

**Comment by houk-ms** [comment]
Error improvement for VM

**Review by arrownj — COMMENTED** [review]
The PR title should start with `{Compute}`, not `{VM}`. And error updates for compute modules are much less than what I expected, are you sure all the work for `Compute` are done?

**Comment by houk-ms** [comment]
> The PR title should start with `{Compute}`, not `{VM}`. And error updates for compute modules are much less than what I expected, are you sure all the work for `Compute` are done?

Of course not, we should do that iteratively. This PR is only for ValueError.

File: src/azure-cli-core/azure/cli/core/azclierror.py [hunk 1]
```diff
@@ -18,6 +18,8 @@ class AzCLIErrorType(Enum):
     # userfaults
     CommandNotFoundError = 'CommandNotFoundError'
     ArgumentParseError = 'ArgumentParseError'
+    UsageError = 'UsageError'
```
Comment (yungezz) [inline src/azure-cli-core/azure/cli/core/azclierror.py:21]: what's usage error means normally? Invalid argument value? It's little too ambiguous, thoughts?

File: src/azure-cli-core/azure/cli/core/azclierror.py [hunk 2]
```diff
@@ -18,6 +18,8 @@ class AzCLIErrorType(Enum):
...
+    ResourceNotFound = 'ResourceNotFound'
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/azclierror.py:22]: I suggest create subclass of each error.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 3]
```diff
@@ -1719,37 +1719,53 @@ def list_vm_images(cmd, image_location=None, publisher_name=None, offer=None, sk
...
+        if len(items) != 4:
+            error_msg = '--urn should be in the format of publisher:offer:sku:version'
+            raise AzCLIError(AzCLIErrorType.UsageError, error_msg)
         publisher, offer, sku, version = urn.split(":")
         if version.lower() == 'latest':
             version = _get_latest_image_version(cmd.cli_ctx, location, publisher, offer, sku)
     elif not publisher or not offer or not sku or not version:
-        raise CLIError(usage_err)
+        error_msg = 'If not using --urn, all of --publisher, --offer, --sku, --version should be specified'
+        raise AzCLIError(AzCLIErrorType.UsageError, error_msg)
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1741]: `raise UsageError(error_msg)`
'UsageError' will be automatically appended.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 4]
```diff
@@ -1719,37 +1719,53 @@ def list_vm_images(cmd, image_location=None, publisher_name=None, offer=None, sk
...
+        items = urn.split(":")
+        if len(items) != 4:
+            error_msg = '--urn should be in the format of publisher:offer:sku:version'
+            raise AzCLIError(AzCLIErrorType.UsageError, error_msg)
         publisher, offer, sku, version = urn.split(":")
         if version.lower() == 'latest':
             version = _get_latest_image_version(cmd.cli_ctx, location, publisher, offer, sku)
     elif not publisher or not offer or not sku or not version:
-        raise CLIError(usage_err)
+        error_msg = 'If not using --urn, all of --publisher, --offer, --sku, --version should be specified'
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1740]: how about make it striaghtforward
```
Pls specify all of (--publisher, --offer, --sku, --version) , or --urn

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 5]
```diff
@@ -1719,37 +1719,53 @@ def list_vm_images(cmd, image_location=None, publisher_name=None, offer=None, sk
...
 def accept_market_ordering_terms(cmd, urn=None, publisher=None, offer=None, plan=None):
+    from azure.cli.core.azclierror import AzCLIErrorType, AzCLIError
     from azure.mgmt.marketplaceordering import MarketplaceOrderingAgreements
 
-    usage_err = 'usage error: --plan STRING --offer STRING --publish STRING |--urn STRING'
     if urn:
         if any([publisher, offer, plan]):
-            raise CLIError(usage_err)
-        publisher, offer, _, _ = urn.split(':')
+            error_msg = 'If using --urn, do not specify any of --plan, --offer, --publish'
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1752]: Could we extra a error message template for this cases; either one param, or a group of splitted params, others just use it by filling in param name?

**Comment by qwordy** [comment]
I suggest create subclass of each error type.
usage: `raise UsageError(error_msg)`
Developer doesn't need to specify error prefix explicitly

**Review by qwordy — APPROVED** [review]
Although I approve, I still suggest use subclass. This is how errors are defined in Python.

**Review by yungezz — APPROVED** [review]

**Review by fengzhou-msft — APPROVED** [review]

**Comment by houk-ms** [comment]
#14987

**Pull request merged**